File: utils/bdmk_wrap.m
function pot = bdmk_warp(nd,ndim,eps,ikernel,beta,ipoly,norder,npbox, ...
                         nboxes,nlevels,ltree,itree,iptr,centers,boxsize,fvals, ...
                         pot)
nhess = ndim*(ndim+1)/2;
%%% based on fvals for all cGTOs, compute phi_k(r')*phi_l(r'), and singular volume integral
nd0 = 1;
disp("=========Start BDMK=======");
disp("nchnk is : " + nd );
for jchnk = 1:nd
  fvalsj = fvals(jchnk,:,:);
  pot0=zeros(nd0,npbox,nboxes);
  grad0=zeros(nd0,ndim,npbox,nboxes);
  hess0=zeros(nd0,nhess,npbox,nboxes);
  tic
  ifpgh=1;
  ifpghtarg=0;
  ntarg = 2;
  targs=zeros(ndim,ntarg); pote=zeros(nd0,ntarg);
  grade=zeros(nd0,ndim,ntarg); hesse=zeros(nd0,nhess,ntarg);
  timeinfo = zeros(20,1);
  [pot0,grad0,hess0,pote,grade,hesse] = ...
    bdmk_mex(nd0,ndim,eps,ikernel,beta,ipoly,norder,npbox, ...
           nboxes,nlevels,ltree,itree,iptr,centers,boxsize,fvalsj, ...
           ifpgh,pot0,grad0,hess0,ntarg,targs, ...
           ifpghtarg,pote,grade,hesse,timeinfo);
  pot(jchnk,:,:) = pot0;
  disp("jchnk is : " + jchnk + " ( total nchnk is : " + nd + " )");
end
disp("=========End BDMK=======");

end

File: utils/f/bdmk_wrap.f90
      subroutine bdmk_wrap(nd,ndim,eps,ikernel,beta,ipoly,norder, &
          npbox,nboxes,nlevels,ltree,itree,iptr,centers,boxsize,  &
          fvals,pot)
      use omp_lib
      implicit none
      real *8 eps, beta
      integer nd, ndim
      integer ikernel, nboxes, nlevels, norder, npbox, ipoly, ltree
      integer itree(ltree), iptr(8)
      real *8 centers(ndim,nboxes)
      real *8 boxsize(0:nlevels)
      real *8 fvals(nd,npbox,nboxes)
      real *8 pot(nd,npbox,nboxes)
      ! 
      integer jchnk, nd0, nhess, ifpgh, ifpghtarg, ntarg
      real *8, allocatable :: fvalsj(:,:,:)
      real *8, allocatable :: pot0(:,:,:),grad0(:,:,:,:),hess0(:,:,:,:)
      real *8, allocatable :: targs(:,:),pote(:,:),grade(:,:,:)
      real *8, allocatable :: hesse(:,:,:),timeinfo(:)
      ! 
      nd0 = 1
      nhess = ndim*(ndim+1)/2
      ifpgh = 1
      ifpghtarg = 0
      ntarg = 2
      ! 
      !$omp parallel default(none) &
      !$omp& shared(nd,ndim,eps,ikernel,beta,ipoly,norder,npbox) & 
      !$omp& shared(nboxes,nlevels,ltree,itree,iptr,centers,boxsize) &
      !$omp& shared(fvals,pot,nd0,nhess,ifpgh,ifpghtarg,ntarg) &
      !$omp& private(jchnk,fvalsj,pot0,grad0,hess0,targs,pote) &
      !$omp& private(grade,hesse,timeinfo)
      ! 
      allocate(fvalsj(nd0, npbox, nboxes))
      allocate(pot0(nd0, npbox, nboxes))
      allocate(grad0(nd0, ndim, npbox, nboxes))
      allocate(hess0(nd0, nhess, npbox, nboxes))
      allocate(targs(ndim, ntarg))
      allocate(pote(nd0, ntarg))
      allocate(grade(nd0, ndim, ntarg))
      allocate(hesse(nd0, nhess, ntarg))
      allocate(timeinfo(20))
      !$omp do 
      do jchnk = 1,nd
        ! print *, 'I am at jchnk = ', jchnk
        ! 
        fvalsj = 0.0d0
        pot0   = 0.0d0
        grad0  = 0.0d0
        hess0  = 0.0d0
        targs  = 0.0d0
        pote   = 0.0d0
        grade  = 0.0d0
        hesse  = 0.0d0
        timeinfo = 0.0d0
        ! 
        fvalsj(1,:,:) = fvals(jchnk, :, :)
        ! 
        call bdmk(nd0,ndim,eps,ikernel,beta,ipoly,norder,npbox, &
          nboxes,nlevels,ltree,itree,iptr,centers,boxsize,fvalsj, &
          ifpgh,pot0,grad0,hess0,ntarg,targs, &
          ifpghtarg,pote,grade,hesse,timeinfo)
        ! 
        pot(jchnk, :, :) = pot0(1, :, :)
        ! print *, 'I am done with jchnk = ', jchnk
      enddo
      !$omp end do
      ! 
      deallocate(fvalsj,pot0,grad0,hess0)
      deallocate(targs,pote,grade,hesse,timeinfo)
      !$omp end parallel
      
      end subroutine bdmk_wrap